### PR TITLE
auto-updater: download versions and archive with binary

### PR DESCRIPTION
**Pull request by mniewrzal**
What: Another part of changes for storage node auto-updater. This PR adds:
* checking current binary version
* downloading suggested version for storage node if the current version is lower than suggested
* downloading archive with binary according to suggested version URL 

For initial testing static version json can be used: https://gist.githubusercontent.com/mniewrzal/77726fe35e26a4c1f474d0640b1d6389/raw/3d1380ad9ae028110ef48a8ea678e6e4cb0238bb/versions.json

Why: We want to have storage node auto-updater (initially for Windows) 
https://storjlabs.atlassian.net/browse/V3-2481

Please describe the tests:
 - Test 1: automatic test for versions comparing, general auto-updater tests will be in last PR
 
Please describe the performance impact: none

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: cmd/auto-updater/main.go [hunk 1]
```diff
@@ -75,6 +154,73 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
...
+	_, err = sync2.Copy(ctx, file, resp.Body)
+	return err
+}
+
+func fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
```
Comment (ifraixedes) [inline cmd/auto-updater/main.go]: It's safer to compare with `ModeType` to ensure that's a regular file and not any other like a symlink, etc.

File: internal/version/version.go [hunk 2]
```diff
@@ -93,6 +93,29 @@ func NewSemVer(v string) (sv SemVer, err error) {
 	return sv, nil
 }
 
+// Compare compare two versions, return -1 if compared version is greater, 0 if equal and 1 if less
```
Comment (ifraixedes) [inline internal/version/version.go]:
```suggestion
// Compare compare two versions, return -1 if compared version is greater, 0 if equal and 1 if less.
```

File: cmd/auto-updater/main.go [hunk 3]
```diff
@@ -75,6 +154,73 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 	return nil
 }
 
+func binaryVersion(location string) (ver version.SemVer, err error) {
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: removed

File: cmd/auto-updater/main.go [hunk 4]
```diff
@@ -75,6 +154,73 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
...
+
+	scanner := bufio.NewScanner(bytes.NewReader(out))
+	for scanner.Scan() {
+		line := scanner.Text()
+		prefix := "Version: "
+		if strings.HasPrefix(line, prefix) {
+			return version.NewSemVer(line[len(prefix):])
+		}
+	}
+	return ver, nil
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: good idea, changed

File: cmd/auto-updater/main.go (hunk 1 above)
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: changed

File: internal/version/version.go (hunk 2 above)
Comment (mniewrzal) [reply internal/version/version.go]: done

File: cmd/auto-updater/main.go [hunk 5]
```diff
@@ -39,6 +51,28 @@ var (
...
+	Storagenode Process `json:"storagenode"`
+}
+
+// Process versions for specific binary
+type Process struct {
+	Minimum   Version `json:"minimum"`
+	Suggested Version `json:"suggested"`
+}
+
+// Version represents version and download url for binary
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: done

File: cmd/auto-updater/main.go [hunk 6]
```diff
@@ -39,6 +51,28 @@ var (
...
+type Response struct {
+	Processes Processes `json:"processes"`
+}
+
+// Processes describes versions for each binary
+type Processes struct {
+	Storagenode Process `json:"storagenode"`
+}
+
+// Process versions for specific binary
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: done

File: cmd/auto-updater/main.go [hunk 7]
```diff
@@ -39,6 +51,28 @@ var (
 	binaryLocation string
 )
 
+// Response response from version server
+type Response struct {
+	Processes Processes `json:"processes"`
+}
+
+// Processes describes versions for each binary
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: done

File: cmd/auto-updater/main.go [hunk 8]
```diff
@@ -39,6 +51,28 @@ var (
 	binaryLocation string
 )
 
+// Response response from version server
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: done

**Comment by mniewrzal** [comment]
> Looks good to me but one thought maybe worth thinking about: Should we be verifying the hash of the downloaded version so that we can detect corrupt downloads and not accidentally try to run a corrupt download?

Yes, there is a plan to do that but we will add this later.

**Pull request merged**